File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int i,j,n=nums.length;
        for(i=n-2;i>=0;i--){
            if(nums[i+1]>nums[i])
                break;
        }
        if(i<0) reverse(nums,i+1);
            else{
                for(j=n-1;j>i;j--){
                    if(nums[j]>nums[i])
                        break;
                }
                swap(nums,i,j);
                reverse(nums,i+1);
            }   
  }
    private void reverse(int[] nums, int start) {
        int i = start, j = nums.length - 1;
        while (i < j) {
            swap(nums, i, j);
            i++;
            j--;
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        n=len(nums)
        if(n==1):
            return 
        idx=0
        for i in range(n-1,-1,-1):
            if(nums[i-1]<nums[i]):# break at last non increasing idx
                idx=i-1 
                break 
                     
        if(idx<0):#last permutation 
            nums.reverse() #nums[::-1]
        else:
            for j in range(n-1,idx,-1):
                if(nums[j]>nums[idx]):
                    break
            nums[j],nums[idx]=nums[idx],nums[j]#swap
            nums[idx+1:]=nums[:idx:-1]#reverse idx+1 to end
                
                
                
            
        
        
        